lib/study_cleanup_tools.rb

```diff
@@ -0,0 +1,174 @@
+      end
+    end
+    study_count
+  end
+
+  # delete all Terra workspaces in a project that do not have a study associated with them in the given env/schema
+  # will ignore any workspaces created by different service accounts/users
+  # by default will only run against test/pentest environment during continuous integration build
+  #
+  # * *params*
+  #   - +project_name+: (String) => Name of Terra billing project (defaults to FireCloudClient::PORTAL_NAMESPACE )
+  #   - +allow_dev_env+: (Boolean) => allow execution of this method in development environment (default: false)
+  #
+  # * *raises*
+  #   - (RuntimeError) => if command cannot be run for failing a security check (hostname, project, environment, etc)
+  #
+  # * *returns*
+  #   - (Integer) => count of workspaces that were deleted
+  def self.delete_all_orphaned_workspaces(project_name = FireCloudClient::PORTAL_NAMESPACE, allow_dev_env: false)
+    halt_on_validation_fail(:permit_environment?, allow_dev_env)
+    halt_on_validation_fail(:permit_hostname?)
+    halt_on_validation_fail(:permit_billing_project?, project_name)
+    halt_on_validation_fail(:is_continuous_integration?)
+    halt_on_validation_fail(:confirm_delete_request?) if !!(allow_dev_env && Rails.env.development?)
+    workspaces = ApplicationController.firecloud_client.workspaces(project_name)
+    workspace_count = 0
+    workspaces.each do |workspace|
+      ws_attr = workspace.dig('workspace')
+      ws_name = ws_attr['name']
+      ws_project = ws_attr['namespace']
+      ws_owner = ws_attr['createdBy']
+      existing_study = Study.find_by(firecloud_project: ws_project, firecloud_workspace: ws_name)
+      if existing_study.present?
+        puts "skipping #{ws_project}/#{ws_name} as it belongs to #{existing_study.accession}"
+        next
+      else
+        begin
+          if service_account_created_workspace?(workspace)
+            print "deleting #{ws_project}/#{ws_name}... "
+            ApplicationController.firecloud_client.delete_workspace(ws_project, ws_name)
+            workspace_count += 1
+            puts "#{ws_project}/#{ws_name} successfully deleted"
+          else
+            puts "skipping #{ws_project}/#{ws_name} because it was created by #{ws_owner}"
+          end
+        rescue => e
+          puts "Unable to delete #{ws_project}/#{ws_name} due to: #{e.class.name} - #{e.message}"
+        end
+      end
+    end
+    workspace_count
+  end
+
+  ##
+  # security safeguards to prevent accidental deletes of production/staging workspaces and studies
+  ##
+
+  # handler to run a security check and raise an error on returning false to halt execution
+  #
+  # * *params*
+  #   - +validation_name+: (String, Symbol) => name of security check to perform
+  #   - +args+: (*Array) => array of arguments for calling method, passed with splat (*)
+  #
+  # * *raises*
+  #   - (RuntimeError) => raise exception if security check returns false to halt parent method execution
```

Comment:
I like what you're doing in centralizing the error handling code and making the check methods pure boolean logic.  I do think the structure of passing validation methods around as symbols and arglists makes the code hard to read/trace, and is generally something that should only be done if you have to.

How about making this a method that takes a block instead?  So the calling code would be
```halt_on_validation_fail { permit_hostname? }```

and this method's implementation would essentially be
```
  result = yield
  if result != true
     raise RuntimeError
  end
```

Also, I'd rename this to ```raise_exception_unless_true``` to be more explicit about its operation

Reply:
That's a great suggestion - I like the idea of invoking the method directly rather than calling `self.send(method, *args).  It definitely is much more explicit.

Reply:
32c1bc3